Side by side diffs

rafeca opened the pull request:
## Description

This is an initial implementation of side by side diffs on GitHub Desktop.

## UX considerations

I've been heavily inspired on the dotcom design/UX to implement this feature, trying to mimic it behaviour as much as possible. This includes a few considerations:

* We only highlight the differences on modified lines when the number of added and deleted lines in a hunk is the same. As an example, in the next screenshot the first hunk has diff highlighting while the second one does not have it:
  ![image](https://user-images.githubusercontent.com/408035/92018707-c6b56080-ed55-11ea-8c27-ad5560e54889.png)
* I've tried to follow as closely as possible the colors/background/borders of dotcom for the different elements (line number boxes, empty spaces, hunk headers, etc) on light mode. Since there's no dark mode in GitHub, I had to go a bit on my own there 🙃 .

In some cases, in order to reuse already existing logic in Desktop, the behaviour is a bit different than in GitHub.com. The most visible instance of this is differences highlighting, where the logic that we use in Desktop differs a bit from the dotcom one (I think that the one in GitHub.com is slightly better, but I don't consider it a big issue to worry about for now):

Desktop:
![image](https://user-images.githubusercontent.com/408035/92019147-725eb080-ed56-11ea-92f3-cf52cfa908bf.png)

GitHub.com:
![image](https://user-images.githubusercontent.com/408035/92019108-5fe47700-ed56-11ea-97d7-3775d83794b7.png)

### Partial committing

In order to have a nicer interface for partial committing, I've decided to move the line numbers of the left pane to the right, this way we have both columns of line numbers together.

This is quite unorthodox, but it solves the problem of selecting multiple lines by drag/drop plus it allows us to use the spacing between the two columns as a gutter that allows selecting the whole hunk (in a kind of similar way as it's done for unified diffs).

I think I like this approach, but I would love to have a more conventional way to solve it if it exists (it's also strange to have the line numbers in the middle in one view and on the left on another). Any suggestion/idea that I haven't considered is welcome here :) 

![demo](https://user-images.githubusercontent.com/408035/92019632-3a0ba200-ed57-11ea-803d-7dd717f226c2.gif)


## Technical considerations

The meaty part of the implementation of this feature is in the `getDiffRowsFromHunk()` method: this function converts a `DiffHunk` object (that contains the different lines of the diff for that hunk) into an array of `DiffRow` objects (which is a new object that represents a line that gets displayed on screen).

This conversion is done by merging groups of consecutive deleted/added lines together in the same `DiffRow` so they can get displayed side by side. Other types of lines are directly converted from a `DiffLine` object to a `DiffRow` object.

Ideally, I'd like to not have `DiffRow` objects, but I found that by creating them with a much more specific TypeScript types than `DiffLine`, the rendering of rows (done in `<SideBySideDiffRow/>` gets quite simpler. A follow-up could be to make the `DiffLine` objects more strict in terms of typing (no nullable fields) which would allow to remove the conversion from `DiffRow` to `DiffLine` objects.

Not using CodeMirror allows us to have more control on the rendering, which makes it much easier to handle click interactions on the lines (e.g we can pass a hunk identifier to each row so we can easily know which hunk to select when clicking on the hunk handler, instead of having to calculate that based on the line clicked).

Also, I could appreciate a performance improvement when quickly scrolling through long diffs: it's quite possible that the overhead that CodeMirror adds is superior to the overhead of React (this is more appreciable when building in prod mode, since in dev mode side by side diffs are penalized by React debugging logic):

Unified diffs (48ms spent in scroll handler):
<img width="1680" alt="Screen Shot 2020-09-02 at 20 26 05" src="https://user-images.githubusercontent.com/408035/92021775-8efce780-ed5a-11ea-98bd-9578875ed5b0.png">

Side By Side diffs (14ms spent in scroll handler):
<img width="1680" alt="Screen Shot 2020-09-02 at 20 26 48" src="https://user-images.githubusercontent.com/408035/92021845-ac31b600-ed5a-11ea-98d1-6d146d08b905.png">



## Next steps

There are a few things that I haven't implemented yet:

1. Auto-scroll the screen when moving the cursor up/down while selecting changes to allow selecting lines that are not shown in the view.

A future next step that I think can be done once/if this PR is merged is to implement unified diffs re-using this same logic. I've taken a quick spike on it and the implementation is relatively straightforward (I only had to change one line in `<SideBySideDiff />` to prevent creating modified rows, but `<SideBySideDiffRow />` and the css needed a few more changes to adjust the layout accordingly).

With this, we could use split diffs to stress test this diff viewer that does not have CodeMirror, and once is stable enough we could enable it for unified diffs behind a feature flag to end up replacing completely the current diff viewer.

## Screenshots

![image](https://user-images.githubusercontent.com/408035/92016802-278f6980-ed53-11ea-96b3-5dd69dd71d11.png)

![image](https://user-images.githubusercontent.com/408035/92016851-34ac5880-ed53-11ea-83b8-5b95b5fcb08a.png)

![image](https://user-images.githubusercontent.com/408035/92016929-50176380-ed53-11ea-964b-c90f0472247e.png)


## Release notes


Notes: [New] Side by side diffs

rafeca committed:
WIP: unified diff

rafeca committed:
WIP commit css

rafeca committed:
WIP: syntax hightlighting

rafeca committed:
big refactor to fix weird issues

rafeca committed:
Move methods around

rafeca committed:
Styling tweaks

rafeca committed:
Moar css tweaks

rafeca committed:
Fix syntax highlighting

rafeca committed:
Rename to unified diff

rafeca committed:
Moar cleanup

rafeca committed:
Allow to select diff text

rafeca committed:
Extract logic outside

rafeca committed:
Moar cleanup

rafeca committed:
Generate unique keys in rows

rafeca committed:
Add setting to show split diffs

rafeca committed:
Copy-pasta works fine now

rafeca committed:
Use more official diff-match-patch npm package

rafeca committed:
Get rid of diff-match-patch in favor of our custom logic

rafeca committed:
WIP: Tweak css styles

rafeca committed:
WIP: do not wrap all chars in span elements

rafeca committed:
Make syntax highlighting logic more robust

rafeca committed:
More robustness

rafeca committed:
Moar refactor

rafeca committed:
Fix color of intra line markings

rafeca committed:
WIP: Partial commiting!!!

rafeca committed:
Tweak text to enable split view

rafeca committed:
WIP: virtualization!!!

rafeca committed:
Total refactor letrol

rafeca committed:
SUPER_WIP: hunk selection

rafeca committed:
Some simplification to the CSS styles

rafeca committed:
Moar docs

rafeca committed:
More more more comments

rafeca committed:
Tweak logic to calculate modified lines

rafeca committed:
Fix weird things

rafeca committed:
cleanup css

rafeca committed:
Moar cleanup

rafeca committed:
Simplify data structures

rafeca committed:
remove console.log

rafeca committed:
Simplify SideBySideDiffRow component

rafeca committed:
Moar simplification

rafeca committed:
moar simplification

rafeca committed:
clean clean clean

rafeca committed:
move stuff to helpers

rafeca committed:
Add no newline indicator to side by side diffs

rafeca committed:
Do not handle hunk hovering when selecting lines

rafeca committed:
Add discard lines functionality

rafeca committed:
Avoid unneeded renders of rows

rafeca committed:
Better hovering of hunks

rafeca committed:
Better naming for hunkid

rafeca committed:
Add docs

rafeca committed:
Moar docs

rafeca committed:
Do not allow to select unselectable diffs

rafeca committed:
Add feature flag

rafeca committed:
Fix eslint error

rafeca committed:
Clear heights cache when changing diffs

Co-authored-by: Markus Olsson <j.markus.olsson@gmail.com>

rafeca committed:
Wrap by letters on side by side diffs

Co-authored-by: Markus Olsson <j.markus.olsson@gmail.com>

rafeca committed:
Revert change in gutter background color

Co-authored-by: Markus Olsson <j.markus.olsson@gmail.com>

rafeca committed:
Fix issue with syntax highlighting

rafeca committed:
Try to fix flickering in size caused by scrollbars in Windows

say25 commented:
This seems like it would collide with https://github.com/desktop/desktop/pull/6818 :( I don't mind rebasing that branch after this work is done.

tierninho commented:
Some notes while testing this:

- Currently, the History tab has the flag to toggle this feature on/off. It seems like it should be on the Changes tab too for obvious reasons.
- Should we add some minimal limitation on how narrow you can make this pane, else we end up with this?
![Screen Shot 2020-09-22 at 9 14 19 AM](https://user-images.githubusercontent.com/14828183/93932340-bc82e400-fcbb-11ea-9055-8ca67633458e.png)
- In the split mode, under Changes, I cannot discard multiple lines like so:
![Screen Shot 2020-09-22 at 10 26 21 AM](https://user-images.githubusercontent.com/14828183/93933981-31571d80-fcbe-11ea-8f41-f0225d77263b.png)

niik reviewed (changes requested):
I'm slowly working my way through this. Overall I think it's fantastic as I've already said in our internal review of this. I'll have some more nits and questions but for now I happened to note that the side by side diffs doesn't contain the diff line type indicators (i.e. `+`, `-`, or ` `).

### Side by side
![image](https://user-images.githubusercontent.com/634063/94035446-5bdece80-fdc3-11ea-9033-c800ab23c54d.png)

### GitHub.com
![image](https://user-images.githubusercontent.com/634063/94035556-84ff5f00-fdc3-11ea-91ce-548b8d14a70a.png)

I think this is because the side by side diff component uses the `.content` property of DiffLine

https://github.com/desktop/desktop/blob/d0b745a9931fb474f240e1fd4edecf378a0696df/app/src/ui/diff/side-by-side-diff.tsx#L579

Which explicitly omits the commit line type marker

https://github.com/desktop/desktop/blob/d0b745a9931fb474f240e1fd4edecf378a0696df/app/src/models/diff/diff-line.ts#L33-L36

rafeca commented:
> Currently, the History tab has the flag to toggle this feature on/off. It seems like it should be on the Changes tab too for obvious reasons.

Yup, completely agree. I can add the toggle on the right of the topbar in a similar way that @say25 did in https://github.com/desktop/desktop/pull/6818 for the whitespace.

Something worth discussing is how we're going to surface this whole new diff viewer... I think we could have it as some kind of experimental feature (and only show the toggle then when the "experimental diff viewer" is enabled), but maybe there are better options.

> Should we add some minimal limitation on how narrow you can make this pane, else we end up with this?

Good point! I think this would be nice to have, specially since the side by side mode needs more horizontal real estate, but I don't think it's a blocker (given the fact that we have a similar problem currently with normal diffs).

> In the split mode, under Changes, I cannot discard multiple lines like so:

I'm not able to reproduce it, when clicking on the gutter between the line numbers I'm able to discard them (maybe it's confusing that the gutter is in the middle?):

![image](https://user-images.githubusercontent.com/408035/94056114-298e9a80-fdde-11ea-90a9-77b0e05057f0.png)

> I happened to note that the side by side diffs doesn't contain the diff line type indicators (i.e. +, -, or ).

Good catch! I'm thinking about adding them manually so I can make them non-selectable by the cursor (in a similar way that dotcom does).

tierninho commented:
> I'm not able to reproduce it, when clicking on the gutter between the line numbers I'm able to discard them (maybe it's confusing that the gutter is in the middle?):

Ah, I see it now. Thanks for the reminder.

rafeca committed:
Add bottom border on diff viewer

rafeca committed:
Add border between columns to readonly diffs

rafeca committed:
Calculate diff tokens in parent component

This will help implement unified diffs

rafeca committed:
Better comments

rafeca committed:
Codestyle tweaks

rafeca committed:
Use new diff viewer for unified diffs

rafeca committed:
Merge branch 'development' into hack-unified-diff

rafeca committed:
Add -/+ symbols before lines

rafeca commented:
> I happened to note that the side by side diffs doesn't contain the diff line type indicators (i.e. +, -, or ).

I've added the add/delete indicators in https://github.com/desktop/desktop/commit/2e10ce4f7dcae85f57bab0e0cd0fc36fa13ff336, which is not part of this PR because I did this on top of the `diff-engine-everywhere` branch which changes a bit the CSS of diffs.

I can merge that branch into this one if this makes the review easier.

cprovatas commented:
So Exciting!!!! Thank you @rafeca!

JosephTLyons commented:
Extremely stoked for this.  I can't test this myself but I'm optimistically hoping it gets added in!

tierninho commented:
> I can merge that branch into this one if this makes the review easier.

At least for me, 👍 

I noticed the gutters are half the normal width when split view in enabled. Very hard to click that thin line -- they should both be doubled in size to match current gutters.

![Screen Shot 2020-09-24 at 2 40 44 PM](https://user-images.githubusercontent.com/14828183/94214146-7e7cf000-fe74-11ea-88cb-7f9e4a07be87.png)

Optional: We could add a right-click context menu in the diff header (on the Changes view side) that enables the user to turn on/off split view. This would remove the need for an awkward checkbox. UI for History tab remains as-is.

rafeca committed:
Better text selection

rafeca commented:
Thanks for the feedback @tierninho !

> I noticed the gutters are half the normal width when split view in enabled. Very hard to click that thin line -- they should both be doubled in size to match current gutters.

Some of the gutters of split diffs are 4px wide (e.g the one in the top of the image you pasted) while some of them are 2px wide. I'm going to try to make all of them 4px wide which is also the width of the current gutters.

Regarding clicking on gutters, the click area of them is much larger (20x20 pixels), which is similar to the current behaviour:

![image](https://user-images.githubusercontent.com/408035/94249621-711d3f80-ff20-11ea-9858-d0b1babb00d7.png)

rafeca committed:
Add context menu for copying text

rafeca committed:
Improve discard line labels

rafeca committed:
Show default cursor when hovering in the hunk information

niik commented:
> I can merge that branch into this one if this makes the review easier.

I took a quick look at the diff of that branch and in general I'd be in favor of pulling it into this but it appeared to me as if that branch made the codemirror-based diff viewer obsolete in its entirety. I think what I'd like to see here is a feature flag to enable the side by side diffs like we have in this PR and then an additional feature flag for whether to use the new diff engine for unified diffs.

It's your call if you want to update that branch with those changes or only port the CSS for the line type indicators over here

rafeca committed:
Improve gutters in selectable side by side diffs

rafeca committed:
Add back enableSideBySideDiffs() flag

We now have two flags: enableSideBySideDiffs() which displays the checkbox to enable side by side diffs and enableExperimentalDiffViewer() which will make unified diffs use the same logic as side by side diffs (which makes the codemirror-based diff viewer obsolete).

rafeca committed:
Add checkbox to enable side by side diffs in the changes tab

rafeca commented:
> I took a quick look at the diff of that branch and in general I'd be in favor of pulling it into this but it appeared to me as if that branch made the codemirror-based diff viewer obsolete in its entirety. I think what I'd like to see here is a feature flag to enable the side by side diffs like we have in this PR and then an additional feature flag for whether to use the new diff engine for unified diffs.

👍 done that and merged that branch into this one!

Some stuff I improved since yesterday:

- Improved context menus to make them look the same as the current diff viewer (same labels to discard lines and added a context menu to copy text).
- Added the +/- signs next to added/removed lines.
- Improved the gutters to select lines/hunk for partial committing (now they're all 4px wide):
    ![image](https://user-images.githubusercontent.com/408035/94275904-cfa9e400-ff47-11ea-8627-3f8636146a7c.png)
    ![image](https://user-images.githubusercontent.com/408035/94275950-de909680-ff47-11ea-8661-1121e0c9a7c8.png)
- Added a simple checkbox to enable the side by side mode in the topbar of the changes tab (probably needs redesign, but it can come later).

rafeca committed:
Avoid using Array.prototype.shift()

Turns out O(n^2) operations can have a huge impact in performance 😅

This change reduces the time to load a 100k line from 20s to 200ms in my machine.

niik commented on app/styles/ui/_side-by-side-diff.scss:
```diff
@@ -0,0 +1,284 @@
+      }
+    }
+
+    &.deleted .before {
+      width: calc(50% + 4px);
+
+      .line-number {
+        border-right-width: 4px;
+      }
+    }
+
+    &.added .after {
+      width: calc(50% + 4px);
+
+      .line-number {
+        border-left-width: 4px;
+      }
+    }
```
These changes seem to have caused a 1px misalignment

![image](https://user-images.githubusercontent.com/634063/94442593-7094db00-01a4-11eb-924e-13cdc6eac700.png)

rafeca committed:
Merge branch 'development' into hack-unified-diff

rafeca committed:
Do not calculate tokens when getting the whole list of rows

This improves significantly the performance of selecting partial commits on large diffs

rafeca committed:
Improve selection of rows in side by side diffs

rafeca committed:
FIx missalignment of columns

rafeca commented:
> These changes seem to have caused a 1px misalignment

Good catch! just fixed it!

I've also did what we talked about the small refactor to the logic that calculates the columns data which has allowed me to improve the selection logic in split diffs to be more similar to selecting by drawing:

![demo](https://user-images.githubusercontent.com/408035/94556903-22dca900-025e-11eb-81a0-fba6eb5e4b35.gif)

niik commented on app/src/lib/equality.ts:
```diff
@@ -77,3 +77,33 @@ export function arrayEquals<T>(x: ReadonlyArray<T>, y: ReadonlyArray<T>) {
+      return false
+    }
+  }
+
+  for (const key of y.keys()) {
+    if (!x.has(key)) {
+      return false
+    }
+  }
```
I doubt it has a practical performance impact but isn't this redundant given that we short-circuit if the size differs?  Shouldn't it be enough to just check that all the keys of `x` exists in `y`?

```suggestion

  for (const key of y.keys()) {
    if (!x.has(key)) {
      return false
    }
  }
```

rafeca replied:
Oh yes you're right!

niik commented on app/src/ui/changes/changed-file-details.tsx:
```diff
@@ -25,6 +32,14 @@ export class ChangedFileDetails extends React.Component<
         <PathLabel path={this.props.path} status={this.props.status} />
         {this.renderDecorator()}
 
+        <Checkbox
```
Can we wrap this in a `enableSideBySideDiffs` feature conditional?

niik commented on app/src/ui/diff/index.tsx:
```diff
@@ -238,6 +243,22 @@ export class Diff extends React.Component<IDiffProps, IDiffState> {
   }
 
   private renderTextDiff(diff: ITextDiff) {
+    if (enableExperimentalDiffViewer() || this.props.showSideBySideDiff) {
```
Should this be `(enableExperimentalDiffViewer() || (enableSideBySideDiffs() && this.props.showSideBySideDiff))`?

rafeca replied:
It should be equivalent since it should not be possible to have `showSideBySideDiff` to true when `enableSideBySideDiffs()` returns false, but yes adding a double check won't hurt us

niik commented on app/src/ui/diff/side-by-side-diff.tsx:
```diff
@@ -0,0 +1,967 @@
+      return
+    }
+
+    // We need to use the event target since the current target will
+    // always point to the container.
+    const target = event.target as HTMLDivElement
+
+    const isSelectingBeforeText = target.closest('.before')
+    const isSelectingAfterText = target.closest('.after')
```
This makes me a little uneasy. We don't know that the thing the  user happened to click on is a `div` do we? And it doesn't seem like we care either, we just want the `.closest` function. Could we do something like this instead?

```suggestion
    if (!(event.target instanceof HTMLElement)) {
      return
    }

    // We need to use the event target since the current target will
    // always point to the container.
    const isSelectingBeforeText = event.target.closest('.before')
    const isSelectingAfterText = event.target.closest('.after')
```

niik commented on app/src/ui/diff/side-by-side-diff.tsx:
```diff
@@ -0,0 +1,967 @@
+      },
+    })
+
+    document.addEventListener('mouseup', this.onEndSelection, { once: true })
```
I know that there's a `once` in there but could we add a corresponding `removeEventListener` in `componentWillUnmount` anyway? I've chased enough weird errors due to use-after-unmount issues in the past.

niik reviewed (changes requested):
I know we're coming right up on when you need to stop working on this @rafeca so don't feel like you have to address these pieces of feedback, I can do that. But if you see a faulty assumption from me I'd appreciate it if you pointed it out

niik committed:
Update app/src/ui/diff/side-by-side-diff.tsx

niik committed:
Ensure we unsubscribe on unmount

niik committed:
Remove redundant key comparison

niik committed:
Move changes view side by side toggle behind feature flag

niik committed:
Strict feature flag

niik committed:
Merge branch 'development' into hack-unified-diff

niik committed:
Remove focus outline from side by side diff grid

It  gets partially overlapped by the rows anyway

niik committed:
Revert unified border

This caused problems in different zoom levels

niik commented:
> * Improved the gutters to select lines/hunk for partial committing (now they're all 4px wide):

I've had to revert this change (i.e https://github.com/desktop/desktop/pull/10617/commits/3c32a6a7c2f910f21f03c61edd80ace894c1946e). While https://github.com/desktop/desktop/pull/10617/commits/d38716d8ad87a79493b327ce63297a8e229d9012 fixed the initial misalignment there were still problems with misaligned columns in other zoom levels. I figured it's not worth holding up this entire PR for and we can take a look at it in a separate PR.

niik committed:
|| !== ??

niik committed:
Readonly arrays are great

niik committed:
:art: Minor style tweaks

niik committed:
:art: cleanup

niik committed:
:art: cleanup

niik committed:
:art: extract inline type

niik committed:
Avoid double iteration of addedDeletedLines

niik reviewed (approved):
This is sufficiently isolated and tested that I'm comfortable bringing it to development. Once #10756 has merged we'll see about releasing this to beta.

Absolutely stellar work here @rafeca.

LeonardLaszlo commented:
When are we going to see this in the official release?

niik commented:
> When are we going to see this in the official release?

We're hoping to ship this in the next production release in early to mid November

victor871129 commented:
Two features that Sublime Merge has and Github Desktop does not was side-by-side diff and viewing all the files at once, modified files and staged files, all files at once. Awesome that you implemented side-by-side diff. Can you implement a setting to view all the modified-staged files at once?

MarcelloTheArcane commented:
This is fantastic!  If it's a new file, it would make sense to show the old single diff, do you think?

![image](https://user-images.githubusercontent.com/21159570/99523500-0f60db00-298f-11eb-83c8-96aa9ceb36d3.png)

JosephTLyons commented:
> This is fantastic!  If it's a new file, it would make sense to show the old single diff, do you think?
> 
> 
> 
> ![image](https://user-images.githubusercontent.com/21159570/99523500-0f60db00-298f-11eb-83c8-96aa9ceb36d3.png)
> 
> 

I think that makes sense.  I suppose it would imply that if it were a single file delete, it would show the old single file view as well?

MarcelloTheArcane commented:
@JosephTLyons  Yep, I forgot deleting files too 😄

ariooo100 reviewed (commented):
delete

mohamaddartomi reviewed (commented):
holle

killmen136 reviewed (commented):
Ed
